File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    static LinkedList<Integer> q = new LinkedList<>();
    static List<Integer>[] graph = new ArrayList[201];
    static boolean[] vis = new boolean[201];
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        for(int i=0;i<n;i++){
            graph[i] = new ArrayList<>();
        }
        
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(i==j) continue;
                if(computers[i][j] == 1){
                    graph[i].add(j);
                }
            }
        }
        
        for(int i=0;i<n;i++){
            if(vis[i]) continue;
            answer++;
            
            q.offer(i);
            vis[i] = true;
            
            while(!q.isEmpty()){
                int start = q.poll();
                
                for(int j=0; j<graph[start].size();j++){
                    Integer num = graph[start].get(j);
                    if(vis[num]) continue;
                    q.offer(num);
                    vis[num] = true;
                }
            }
            
        }
        
        
        return answer;
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.kt
import java.util.*

val networks = Array(201){mutableListOf<Int>()}
val vis = Array(200){false}
var cnt = 0

class Solution {
    fun solution(n: Int, computers: Array<IntArray>): Int {
        var answer = 0
        
        for(i in computers.indices){
            for(j in computers[i].indices){
                if(i==j) continue
                if(computers[i][j]==1){
                    networks[i].add(j)
                }
            }
        }
        
        bfs(n)
        
        
        return cnt
    }
    
    fun bfs(n: Int){
        val q = LinkedList<Int>()
        for(i in 0..n-1){
            if(vis[i]) continue
            q.offer(i)
            vis[i] = true
            cnt++
            while(!q.isEmpty()){
                val cur = q.poll()
                networks[cur].forEach{
                    if(vis[it]) return@forEach
                    q.offer(it)
                    vis[it] = true
                }
            }
        }
    }
}